### ISSUE TITLE
Avoid redundant calls to Proxy.upgrade()

### ISSUE THREAD

**Issue by ajsantander** [OpenZeppelin/openzeppelin-upgrades#289]
I'm trying out the hardhat plugin for upgrades, and I noticed that the plugin is smart enough to know when a new implementation doesn't need to be deployed when calling `upgradeProxy`, but it does always call the proxy's `upgrade` function. It could also be smart here and avoid a redundant function call that will only re-set the existing values.

So, essentially, calling `upgradeProxy` on a proxy, with a local implementation that did not change, should be a no-op.

**Comment by frangio** [comment]
Thanks for reporting! I'll close this as a duplicate of https://github.com/OpenZeppelin/openzeppelin-upgrades/issues/89, and bump the priority for the issue.

Can you share a little bit about why this is important to you? I'd like to understand what kind of workflow causes this to be an issue.

**Comment by ajsantander** [comment]
Sure.

TL;DR: Automation for code bases with potentially hundreds of proxies..

In Synthetix, we have a pretty complex deploy script. It's quite smart in the sense that it will not do anything if it doesn't need to. I.e. it will compare deployed implementations with local code, on-chain storage values with expected parameters, etc. If a value is suppose to be 42, it will read the value and figure out if the tx can be avoided. Similarly, if compiling the local source of a contract produces the same bytecode as a proxies current implementation, it will figure out if an upgrade can be skipped.

I'm hoping to use the upgrades plugin in a similar manner. I'd call `upgradeProxy` on all proxies, and it would do nothing if nothing is required. For now, I'm playing with `prepareUpgrade`, and checking if it returns a known, already deployed, implementation. If so, it means that I don't need to call `upgradeProxy`.

**Comment by frangio** [comment]
That makes a lot of sense. Thanks for the explanation! Closing to continue on #89.
